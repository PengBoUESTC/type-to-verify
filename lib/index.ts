import {
  Project,
  type SourceFile,
  SyntaxKind,
  type Type,
  ts,
  SymbolFlags,
  type InterfaceDeclaration,
  type TypeAliasDeclaration,
  type EnumDeclaration,
} from 'ts-morph'
import { cwd } from 'node:process'
import { join } from 'node:path'

interface Options {
  /** cache TypeToValue instance or not */
  cache?: boolean
  /** file path need to parse */
  sourceFilePath: string
  /** ts confing file path */
  tsConfigFilePath?: string
}

export class TypeToVerify {
  project: Project
  constructor(options: Options) {
    this.project = this.init(options)
  }

  init(options: Options) {
    const { sourceFilePath, tsConfigFilePath = join(cwd(), 'tsconfig.json') } =
      options
    const project = new Project({
      tsConfigFilePath,
    })
    project.addSourceFilesAtPaths(sourceFilePath)

    return project
  }

  genLiteralValue(type: Type<ts.Type>) {
    if (type.isBooleanLiteral()) {
      // 获取布尔字面量的值
      return type.getText() === 'true'
    }
    return type.getLiteralValue()
  }

  genEnum(enumDeclaration?: EnumDeclaration) {
    if (!enumDeclaration) return []
    const members = enumDeclaration.getMembers()
    return members
  }

  generateValidator(type: Type, key = 'value'): string {
    if (type.isUndefined() || type.isVoid()) {
      return `${key} === undefined`
    }

    if (type.isNull()) {
      return `${key} === null`
    }

    if (type.isLiteral()) {
      return `${key} === ${this.genLiteralValue(type)}`
    }
    if (type.isString()) {
      return `typeof ${key} === "string"`
    }
    if (type.isNumber()) {
      return `typeof ${key} === "number"`
    }
    if (type.isBoolean()) {
      return `typeof ${key} === "boolean"`
    }
    if (type.isEnum()) {
      return `${this.genEnum(
        type
          .getSymbol()
          ?.getDeclarations()[0]
          .asKindOrThrow(SyntaxKind.EnumDeclaration),
      )}.includes(${key})`
    }
    if (type.isArray()) {
      const elementType = type.getArrayElementTypeOrThrow()
      return `Array.isArray(${key}) && ${key}.every((element, i) => ${this.generateValidator(elementType, `${key}[i]`)})`
    }
    if (type.isObject()) {
      const properties = type.getProperties()
      const checks = properties.map(prop => {
        const propName = prop.getName()
        const propType = prop.getTypeAtLocation(prop.getValueDeclarationOrThrow())
        const isOptional = prop.hasFlags(SymbolFlags.Optional)
        const check = `${this.generateValidator(propType, `${key}['${propName}']`)}`
        return isOptional
          ? `(!value.hasOwnProperty('${propName}') || ${check})`
          : `value.hasOwnProperty('${propName}') && ${check}`
      })
      return `typeof value === 'object' && value !== null && ${checks.join(' && ')}`
    }

    return 'true' // Fallback for any type
  }

  runBase(
    interfaceDeclaration?: InterfaceDeclaration | TypeAliasDeclaration,
  ) {
    if (!interfaceDeclaration) return 'return true'
    return `return ${this.generateValidator(interfaceDeclaration.getType())};`
  }

  run(path: string, typeValue: string) {
    const sourceFile = this.project?.getSourceFile(path)
    if (!sourceFile) return 'return true'
    return this.runBase(
      sourceFile.getInterface(typeValue) || sourceFile.getTypeAlias(typeValue),
    )
  }
}
