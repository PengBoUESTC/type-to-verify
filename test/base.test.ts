import type { Address } from './dts/Address'

import { TypeToVerify } from '../lib/index'

const typeToVerify = new TypeToVerify({
  sourceFilePath: 'test/**/*.d.ts'
})

describe('base test',  () => {
  const result = typeToVerify.run('test/dts/Address.d.ts', 'Address')
  const address: Address = {
    province: 'ce',
    // provinceName: 1,
    child: [1, 2]
  }
  test('base test', () => {
    expect(result).toMatchSnapshot()
    const verify = new Function('value', result)
    expect(verify.toString()).toMatchSnapshot()
    expect(verify({
      address: ''
    })).toBeFalsy()
    expect(verify(address)).toBeTruthy()
  })
})