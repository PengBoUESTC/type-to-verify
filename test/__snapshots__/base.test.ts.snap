// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`base test base test 1`] = `"return typeof value === 'object' && value !== null && value.hasOwnProperty('province') && typeof value['province'] === "string" && (!value.hasOwnProperty('provinceName') || true) && value.hasOwnProperty('child') && Array.isArray(value['child']) && value['child'].every((element, i) => typeof value['child'][i] === "number");"`;

exports[`base test base test 2`] = `
"function anonymous(value
) {
return typeof value === 'object' && value !== null && value.hasOwnProperty('province') && typeof value['province'] === "string" && (!value.hasOwnProperty('provinceName') || true) && value.hasOwnProperty('child') && Array.isArray(value['child']) && value['child'].every((element, i) => typeof value['child'][i] === "number");
}"
`;
